src/undo.h

```diff
@@ -7,70 +7,90 @@
 #define BITCOIN_UNDO_H
 
 #include "compressor.h" 
+#include "consensus/consensus.h"
 #include "primitives/transaction.h"
 #include "serialize.h"
 
 /** Undo information for a CTxIn
  *
- *  Contains the prevout's CTxOut being spent, and if this was the
- *  last output of the affected transaction, its metadata as well
- *  (coinbase or not, height, transaction version)
+ *  Contains the prevout's CTxOut being spent, and its metadata as well
+ *  (coinbase or not, height). The serialization contains a dummy value of
+ *  zero. This is be compatible with older versions which expect to see
+ *  the transaction version there.
  */
-class CTxInUndo
+class TxInUndoSerializer
 {
-public:
-    CTxOut txout;         // the txout data before being spent
-    bool fCoinBase;       // if the outpoint was the last unspent: whether it belonged to a coinbase
-    bool fCoinStake;      // if the outpoint was the last unspent: whether it belonged to a coinstake
-    unsigned int nHeight; // if the outpoint was the last unspent: its height
-    int nVersion;         // if the outpoint was the last unspent: its version
-
-    CTxInUndo() : txout(), fCoinBase(false), fCoinStake(false), nHeight(0), nVersion(0) {}
-    CTxInUndo(const CTxOut &txoutIn, bool fCoinBaseIn = false, bool fCoinStakeIn = false, unsigned int nHeightIn = 0, int nVersionIn = 0) : txout(txoutIn), fCoinBase(fCoinBaseIn), fCoinStake(fCoinStakeIn), nHeight(nHeightIn), nVersion(nVersionIn) { }
+    const Coin* txout;
 
+public:
     template<typename Stream>
     void Serialize(Stream &s) const {
-/////////////////////////////////////////////////////////// // qtum
-        // ::Serialize(s, VARINT(nHeight*2+(fCoinBase ? 1 : 0)));
-        ::Serialize(s, VARINT(nHeight));
-        ::Serialize(s, fCoinBase);
-        ::Serialize(s, fCoinStake);
-///////////////////////////////////////////////////////////
-        if (nHeight > 0)
-            ::Serialize(s, VARINT(this->nVersion));
-        ::Serialize(s, CTxOutCompressor(REF(txout)));
+        unsigned int nCode = txout->nHeight*4 + (txout->fCoinStake ? 2 : 0) + (txout->fCoinBase ? 1 : 0);
```

Comment:
This can be changed to (txout->nHeight << 2) to be more clear that this is building a bitfield, and potentially be more optimized